core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,653 @@
+    }
+
+
+    @Override
+    public void attach() {
+      dummy.attach();
+    }
+
+    @Override
+    public void detach() {
+      dummy.detach();
+    }
+
+    @Override
+    public boolean isCurrent() {
+      return dummy.isCurrent();
+    }
+
+    /**
+     * Attach this context to the thread and return a {@link java.lang.AutoCloseable} that can be
+     * used with try-with-resource statements to properly {@link #detach} and {@link #cancel}
+     * the context on completion.
+     *
+     * @return a {@link java.io.Closeable} which can be used with try-with-resource blocks.
+     */
+    public Closeable attachAsCloseable() {
+      attach();
+      return new Closeable() {
+        @Override
+        public void close() throws IOException {
+          detachAndCancel(null);
+        }
+      };
+    }
+
+    /**
+     * Cancel this context and optionally provide a cause for the cancellation, all listeners
+     * are notified immediately on the calling thread.
+     *
+     * @return {@code true} if this context cancelled the context and notified listeners,
+     *    {@code false} if the context was already cancelled.
+     */
+    public boolean cancel(@Nullable Throwable cause) {
+      synchronized (this) {
+        if (!cancelled) {
+          cancelled = true;
+          if (scheduledFuture != null) {
+            // If we have a scheduled cancellation pending attempt to cancel it.
+            scheduledFuture.cancel(false);
+            scheduledFuture = null;
+          }
+          this.cause = cause;
+          notifyAndClearListeners();
+          return true;
+        }
+        return false;
+      }
+    }
+
+    /**
+     * Cancel this context and detach it from the current context from the thread and restore the
+     * context that was previously attached to the thread as the 'current' context.
+     *
+     * @throws java.lang.IllegalStateException if this context is not {@link #current()}.
+     */
+    public void detachAndCancel(@Nullable Throwable cause) {
+      try {
+        detach();
+      } finally {
+        cancel(cause);
+      }
+    }
+
+    @Override
+    protected boolean canBeCancelled() {
+      return true;
+    }
+
+    @Override
+    public boolean isCancelled() {
+      synchronized (this) {
+        if (cancelled) {
+          return true;
+        }
+        // Detect cancellation of parent in the case where we have no listeners and
+        // record it.
+        if (super.isCancelled()) {
+          cancel(super.cause());
+          return true;
+        }
+      }
+      return false;
+    }
+
+    @Nullable
+    @Override
+    public Throwable cause() {
+      if (isCancelled()) {
+        return cause;
+      }
+      return null;
+    }
+  }
+
+  /**
+   * A listener notified on context cancellation.
+   */
+  public interface CancellationListener {
+    /**
+     * @param context the newly cancelled context.
+     */
+    public void cancelled(Context context);
```

Comment:
It seems the `context` passed here is more annoying to use since there isn't a `Context.get(Key)`. It seems highly likely that either the `context` will be ignored or the Context will have to be attached. Attaching automatically seems a bit too magical, especially if the application provided an executor that set the `Context` they wanted (since us attaching will override the executor's Context). The listener _could_ get the `cause()`, but that's about the only useful thing it could do with the `Context` directly.